### ISSUE TITLE
Deduplication

### ISSUE THREAD

**Issue by michielbdejong** [federatedbookkeeping/world-ledger#63]
Need both for personal finance insight and for federated timesheets

**Comment by michielbdejong** [comment]
Needed in particular also for https://github.com/pondersource/prejournal/issues/37

**Comment by michielbdejong** [comment]
Also for source documents we may want to refer to line numbers for csv, and to xpath for xml, etc?

**Comment by Angus-McAllister** [comment]
Reflecting on this topic in the light of our conversation on Monday 2022-08-15, _unambiguous identifiers_ for entities in remote systems (i.e. those to which an integration has been created) are certainly needed.  When data are propagated from such systems into those participating in a federation (such as **PreJournal**, **timeld**, and **Tiki**), the federated systems will likely receive the data from more than one source, since the data are essentially being gossiped.  Hence, assuming the data in question are unchanged after leaving their source remote system, the federated system receiving them a second or subsequent time should reject them (note that this doesn't mean rejecting an entire POST request, but just the duplicate data therein - the receiving federated system's behaviour could just be to drop the duplicate data, returning a [207](https://httpstatuses.io/207) response overall, with [201](https://developer.mozilla.org/en-US/docs/web/http/status/201) sub-responses for data new to it, and [409](https://developer.mozilla.org/en-US/docs/web/http/status/409) for duplicates).

This would provide the following benefits:

1. Prevent duplication of data altogether (rather than remedy it after the fact);
2. Enable distributed audit as an element of data governance in a federated context.

As to the format of the remote system's identifier of those entities, a URI would be the optimum approach, as discussed, since first, it would uniquely identify not only the application representing the source of those data, but the specific instance of that application - the system, in this case.  And second, it would also uniquely identify the entity obtained from that system.

**Comment by michielbdejong** [comment]
Right! This is what I meant, yes, there is 1 copy already in the db and a 2nd copy of the same data comes in, as "a duplicate of what we already have" and is thus discarded.

> a URI would be the optimum approach

Yes, and we can use that in the timesheets federation also for the reasons you mention, but we shouldn't demand it in general.

Good point about not using line numbers in source documents, but trying to use some identifier that is at home in the source system. I'll see how to do that where possible, although I can think of situation where it will *not* be possible (i.e. if a CSV file simply doesn't contain any sort of "id" column).
